File: strings/longest-word-in-dictionary-through-deleting/longest-word-in-dictionary-through-deleting.js
/**
 * @param {string} s
 * @param {string[]} d
 * @return {string}
 */
function findLongestWord(s, d) {
  let maxString = '';

  for (const str of d) {
    if (search(s, str)) {
      if (str.length > maxString.length || (str.length === maxString.length && str < maxString)) {
        maxString = str;
      }
    }
  }
  return maxString;
}

/**
 *
 * @param {string} s
 * @param {string} word
 * @return {boolean}
 */
function search(s, word) {
  let i = 0;
  let j = 0;
  let n = s.length;
  let m = word.length;

  while (i < n && j < m) {
    if (s[i] === word[j]) {
      i++;
      j++;
    } else {
      i++;
    }
  }
  return j === m;
}

File: strings/longest-word-in-dictionary-through-deleting/longest-word-in-dictionary-through-deleting.go
package longestwordindictionarythroughdeleting

func findLongestWord(s string, d []string) string {
	maxString := ""

	for _, str := range d {
		if search(s, str) {
			if len(str) > len(maxString) || (len(str) == len(maxString) && str < maxString) {
				maxString = str
			}
		}
	}
	return maxString
}

func search(s string, word string) bool {
	i := 0
	j := 0
	n := len(s)
	m := len(word)

	for i < n && j < m {
		if s[i] == word[j] {
			i++
			j++
		} else {
			i++
		}
	}
	return i == m
}
